### PR TITLE
docs: OEP-67 - Document our best practice tools and tech.

### PR DESCRIPTION
**Pull request by feanil**
## The topic for consideration in this PR
We should have an OEP to help us know which tools and technology we as a community should standardize on.  This OEP will help us track these items and provide a structure place for us to have future conversations when we want to change community standards.

We currently have some of this for the frontend but having something a
bit more standard across the entire ecosystem would be helpful for
making decisions going forward.

**Note for Reviewers**: The tools/standards in the frontend section were simply moved from OEP-11, they are NOT up for re-consideration in this PR.  If they are wrong or need updates, I recommend supporting this OEP and then making a new ADRs to update a recommendation that you feel is incorrect.

### Tasks for After the OEP is Approved

- [x] Update OEP-11 to be superseded
- [x] Move ADRs from OEP-11 to OEP-67
- [x] Generate all the relevant redirects.
- [ ] Add a new overall ADR for Codecov, since technically it moves from being a frontend recommendation to being an overall recommendation.

### CODE REVIEW

**Comment by brian-smith-tcril** [comment]
This is great! It looks like everything from OEP-11 can go under the "Frontend Technology Selection" section as it's all pretty frontend specific.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (brian-smith-tcril) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:117]: Is the idea to have `oeps/best-practices/oep-0067/decisions/thing-we-decided.rst` or `oeps/best-practices/oep-0067/decisions/frontend/thing-we-decided.rst`?

Comment (feanil) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:117]: I had assumed the first one but I'm very open to the second option.  What's the value of the second option in your mind, I see that it provides a bit more separation but I'm not clear on what that separation buys us.

Comment (brian-smith-tcril) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:117]: I'm fine either way. I guess I was thinking the separation would buy us a nested TOC. This was asked out of curiosity not opinion.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (e0d) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:43]: This ends up being pretty repetitive of the Abstract, may not be easily avoidable.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (e0d) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:91]: Should the definition just be a note?

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (e0d) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:299]: Add tech-radar consequences.

**Review by e0d — COMMENTED** [review]
Review of the current draft.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:62]: I like that you kept each of the Rationale sections within the Specification fairly brief. The linked ADRs will contain plenty of detail, I imagine, so to reduce duplication and ease readability, it will be good to just stick to summaries here.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:79]:
```suggestion
#. **Codecov should be used to report code coverage by unit tests**

   **Rationale**: It is important to measure the amount of code covered by our
   automated test suites. By striving for a high level of test coverage, we can
   reduce the number of bugs that can only be found via manual testing, 
   and by using codecov to run coverage in CI,
   contributors are automatically reminded to include tests for any new code.

   The Open edX community has standardized around `Codecov`_ as a coverage reporting service.
```

(Nitpicking this particular entry a bit, but I think this will be relevant for other technology choices too.)

Codecov feels different than, say, React, because it's a proprietary SaaS thing. If the offering starts to get bad, we will need to switch, even if we like the tools that come with the service. I think it's fine to list it alongside React et al, but I want to be clear that this isn't really a tool, it's a service.

And if Codecov does ever get bad, we can continue locally measuring Python code coverage using coverage.py (which might be what Codecov is using under the hood... not sure).

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:64]: Consider ordering the spec so that the import stuff is generally closer to the top. Seeing Codecov as the first item in the spec feels a little off, given how minor of a choice it is in the grand scheme of things. Maybe having the subsection order be [Python, JS, General] would be more conducive to that.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:303]: Yay for consolidation!

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:114]: I remember David Joy saying at one point that Redux wasn't worth the hassle and that MFE devs should feel free to simply not use it.

I'm not sure if I remember that right or whether that statement still applies, I'd need @arbrandes or @brian-smith-tcril 's input.

(I guess this is the sort of thing that will make this OEP useful :)

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:162]:
```suggestion
#. **JavaScript projects should publish a package-lock file**
```

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:201]:
```suggestion
#. **CSS should be generated using SCSS**
```

We want people to use the SCSS syntax and *not* the original Sass syntax, right? If so , let's put that on the first line.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:2]:
```suggestion
OEP-67: Standard Tools and Technologies
#######################################
```

Echoing my other comment here, but consider a title that clarifies that this OEP is the place to pick which tools and technologies we use, not the place to discuss things like coding standards, a11y standard practices, etc.

**Review by kdmccormick — COMMENTED** [review]
Thanks Feanil. This is long overdue. A few comments but I'm very supportive of the overall direction.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (feanil) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst]: I think it's useful to enumerate the specific services here because often many services can do the same job but picking one and listing it is a useful way to reduce fracturing across the ecosystem.  Also, if tomorrow Codecov got bad, we don't currently have a good place to have the conversation about what the alternatives are and why we would pick one over others.  I think this OEP will provide the place for that conversation for both tools and SaaS services we use as a community.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (feanil) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:113]: This is a copy from the current frontend OEP.  I think any corrections should be made as follow-up ADRs and updates to the OEP.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (feanil) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:63]: Thinking more about this, I think it would be fine to Move general to the bottom so the order would be [JS, Python, General]  let me know if others have opinions on this and I can make a decisions at some point based on feedback.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (jmbowman) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:173]: Shouldn't this recommend Renovate instead?  See https://openedx.atlassian.net/wiki/spaces/AC/pages/1841791377/Set+up+Renovate+to+Automate+JavaScript+Dependency+Updates#Notes-on-Alternatives for reasons we started to ditch Greenkeeper.  As with the Redux note above we could update it after initial release, but I'd rather not risk people trying to start switching to Greenkeeper after reading this.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (deborahgu) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:92]: Do we actually standardise on "frontend", one word, no punctuation? It's fine if we do, that's just sufficiently unusual that it will probably trigger various spellcheckers and auto-correct tools.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (deborahgu) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:100]: Is the guidance really MUST on React? Some UIs are going to be so simple that react is overkill. (See above regarding: "new around here"; if this debate has already been had and resolved then I don't need anybody to restart it for me. It's more a question about unexamined assumptions.)

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (brian-smith-tcril) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:92]: As far as I know we do standardize on "frontend." It does seem to trigger spellcheckers, but it also seems to be a very common spelling. For example, [the wiki page for "Frontend and backend"](https://en.wikipedia.org/wiki/Frontend_and_backend) notes:
> the terms frontend and backend (sometimes written as back end or back-end)

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (nedbat) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:36]: As long as we are nitting:
```suggestion
The Open edX Platform is built using many first- and third-party tools and technologies.
```

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (feanil) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:299]: In my mind, the radar might still be useful as a graphical interface for this data.  This might imply that we should add the radar states to the entries for each item in this oep so it can also convey provisional or hold items along with what is currently implicitly adopt items.

I'm curious to hear what others think about this.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (feanil) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:100]: that's a fair point, I'm not sure what the intent of the original authors was here, per your second point, MUST feels like a lot here to me as well but I'm not sure what the current stance should be.  This is probably worth a follow-up conversation in the frontend working group.  @brian-smith-tcril @arbrandes @adamstankiewicz what do you think?

Comment (adamstankiewicz) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:100]: @deborahgu If a UI isn't using React, at least one possible concern is the compatibility with using the Paragon design system and React-based UI component library in terms of brand consistency, etc. across UI projects within Open edX.

Comment (deborahgu) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:100]: yeah, this is a bigger question than can be resolved in this OEP, but eg. right now we're working on a place where we're dealing with the entire Paragon dependency stack to render 100% static pages, and it's not lightweight. It might be fundamentally worth the cost, but the cost is not small.

Comment (arbrandes) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:100]: Yeah, I actually think we're squarely in `SHOULD`-land; even React.  This definitely deserves an dedicated conversation.

**Comment by feanil** [comment]
@bradenmacdonald the review period is complete, I think there are a lot of follow-up updates but as far as I can tell, no disagreement on this OEP existing and the goals it's trying to solve for. Do you have any concerns that I should address because I make the rest of the re-organization changes and land this OEP?

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:48]: Overindented, is rendering as a quotation rn
```suggestion
* It is difficult to know the current best practice for a given area.

* It is unclear how to go about the discussion and decision-making process
      to change a best practice.
```

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:58]:
```suggestion
* Is considered a best practice for the Open edX Platform

* Is relevant to a large subset of the Open edX Platform
```

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:302]: Friendly reminder to go into OEP-11 and make this change.

**Comment by bradenmacdonald** [comment]
@feanil Nope! I don't have any concerns, and I think people seem quite happy with this change. (And eager to make the follow-up updates, as it seems like quite a few things have gotten outdated from the original.)

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:10]:
```suggestion
     - :doc:`OEP-67 <oep-0067-bp-tools-and-technology>`
```

**Review by kdmccormick — APPROVED** [review]
LGTM, pending a couple formatting fixes.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (kdmccormick) [reply oeps/best-practices/oep-0067-bp-tools-and-technology.rst:63]: @feanil I do like that idea, but it's not blocking for me.

**Comment by feanil** [comment]
@bradenmacdonald good call, I'll make a new top-level issue and collect all the known updates we need into that for now.

**Comment by feanil** [comment]
@bradenmacdonald  I've created https://github.com/openedx/open-edx-proposals/issues/531 to track the current known updates that we need.  Let me know if you think there's anything else that needs to be done before we land this.

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (bradenmacdonald) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:79]:
```suggestion
   **Rationale**: As a healthy open source project we need good continuous integration
```

File: oeps/best-practices/oep-0067-bp-tools-and-technology.rst
Comment (bradenmacdonald) [inline oeps/best-practices/oep-0067-bp-tools-and-technology.rst:81]:
```suggestion
   tool should be used to perform Continuous Integration (CI) testing for all Open edX repositories
```

**Review by bradenmacdonald — APPROVED** [review]
@feanil Thanks for making https://github.com/openedx/open-edx-proposals/issues/531 . I think this is ready to go!

**Pull request merged**
